File: web/src/crypto/aead.ts
import { randomBytes } from './bytes';

export const NONCE_LENGTH = 12;
export const KEY_LENGTH = 32;

export interface Sealed {
  ciphertext: Uint8Array;
  nonce: Uint8Array;
}

/**
 * AES-256-GCM through WebCrypto. It is the only symmetric primitive the browser offers
 * natively, and its 12-byte nonce fits the bounds the API already validates.
 */
export async function importKey(raw: Uint8Array): Promise<CryptoKey> {
  if (raw.length !== KEY_LENGTH) throw new Error(`content key must be ${KEY_LENGTH} bytes`);

  return crypto.subtle.importKey('raw', raw as BufferSource, 'AES-GCM', true, ['encrypt', 'decrypt']);
}

export async function generateKey(): Promise<CryptoKey> {
  return crypto.subtle.generateKey({ name: 'AES-GCM', length: 256 }, true, ['encrypt', 'decrypt']);
}

export async function exportKey(key: CryptoKey): Promise<Uint8Array> {
  return new Uint8Array(await crypto.subtle.exportKey('raw', key));
}

export async function encrypt(key: CryptoKey, plaintext: Uint8Array, aad: Uint8Array): Promise<Sealed> {
  const nonce = randomBytes(NONCE_LENGTH);

  const ciphertext = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: nonce as BufferSource, additionalData: aad as BufferSource },
    key,
    plaintext as BufferSource,
  );

  return { ciphertext: new Uint8Array(ciphertext), nonce };
}

/**
 * Fails on any tampering, including a ciphertext moved to a different slot: the caller's
 * additional data binds it to one entity, scope and key version.
 */
export async function decrypt(
  key: CryptoKey,
  sealed: Sealed,
  aad: Uint8Array,
): Promise<Uint8Array> {
  const plaintext = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: sealed.nonce as BufferSource, additionalData: aad as BufferSource },
    key,
    sealed.ciphertext as BufferSource,
  );

  return new Uint8Array(plaintext);
}

File: internal/envelope/aead.go
package envelope

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"fmt"
)

const (
	// NonceLength is what WebCrypto uses for AES-GCM and what the API validates.
	NonceLength = 12
	KeyLength   = 32
)

// Sealed is a ciphertext together with the nonce it was produced under. The two are stored
// in separate columns everywhere, which is why they travel as a pair rather than concatenated.
type Sealed struct {
	Ciphertext []byte
	Nonce      []byte
}

func encrypt(key, plaintext, aad []byte) (Sealed, error) {
	gcm, err := newGCM(key)
	if err != nil {
		return Sealed{}, err
	}

	nonce := make([]byte, NonceLength)
	if _, err := rand.Read(nonce); err != nil {
		return Sealed{}, fmt.Errorf("read nonce: %w", err)
	}

	return Sealed{Ciphertext: gcm.Seal(nil, nonce, plaintext, aad), Nonce: nonce}, nil
}

// decrypt fails on any tampering, a ciphertext moved to a different slot included: the
// caller's additional data binds it to one entity, scope and key version.
func decrypt(key []byte, sealed Sealed, aad []byte) ([]byte, error) {
	gcm, err := newGCM(key)
	if err != nil {
		return nil, err
	}

	if len(sealed.Nonce) != NonceLength {
		return nil, fmt.Errorf("nonce must be %d bytes, got %d", NonceLength, len(sealed.Nonce))
	}

	plaintext, err := gcm.Open(nil, sealed.Nonce, sealed.Ciphertext, aad)
	if err != nil {
		return nil, fmt.Errorf("open: %w", err)
	}

	return plaintext, nil
}

func newGCM(key []byte) (cipher.AEAD, error) {
	if len(key) != KeyLength {
		return nil, fmt.Errorf("content key must be %d bytes, got %d", KeyLength, len(key))
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("new cipher: %w", err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("new gcm: %w", err)
	}

	return gcm, nil
}
